README
usermon - User Monitor
Copyright 2003-2021 Fabrice Colin <fabrice dot colin at gmail dot com>
https://github.com/FabriceColin/usermon

This is a plugin for the Xfce panel, version 4.12 or higher.
It displays at all times how many users are currently logged in,
issues a desktop notification when a user logs in and subsequently
when one logs out.

Requirements
============

libxfce4ui-2		>= 4.12.0
libxfce4panel-2.0	>= 4.12.0
libnotify		>= 0.7.0
a session management system (or display manager) that updates utmp

Settings
========

Critical Number of Users sets the threshold starting at which
critical notifications are shown.
Alarm Period (in seconds) defines how often usermon should check
utmp for new users.

Acknowledgements
================

Version 1.4.0 was mostly developed on a PowerBook5,4 running the
excellent Void Linux PPC.

History
=======

The current usermon is a Xfce port of a previous version targeting
Jonas Aaberg's General Applet Interface library [1], developed in
2003 and 2005 and never subsequently published.

[1] https://gai.sourceforge.net/library.html



NEWS
15/08/2021 : v1.4.1
===================
 tweaked the preferences dialog
 check for the right version of Xfce

19/09/2020 : v1.4.0
===================
 initial port to the Xfce Panel

27/08/2005 : v1.3.3
===================
 compatible with GAI 0.5.10

25/09/2004 : v1.3.2
===================
 minor changes
 requires GAI 0.5.8 (fixes a couple of memory leaks)

03/06/2004 : v1.3.0
===================
 use default GTK font, as detected by GAI (thanks Jonas !)
 requires GAI 0.5.6

11/04/2004 : v1.2.0
===================
 made font family/size configurable
 refresh background more often to fix "right-hand side border" bug
 added loud speaker to icon to represent an alarm
 relabeled "Maximum number of users" to "Critical..." since that's what it is really
 finally realized that "ignore root" & "ignore current user" should take effect immediately doh ! :-)
 make use of new preferences window generator
 use asynchronous gai_exec() instead of system()
 requires GAI 0.5.3
 switched build system to autoconf, based on the templates that come with GAI 0.5.3

30/01/2004 : v1.1.0
===================
 added menu to applet that lets one disable alarms for all current users for the lifetime of the applet
 requires GAI 0.5

9/10/2003 : v1.0.0
===================
 initial version



autogen.sh
#!/bin/sh
#
# Copyright 2003-2020 Fabrice Colin
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#

(type xdt-autogen) >/dev/null 2>&1 || {
  cat >&2 <<EOF
autogen.sh: You don't seem to have the Xfce development tools installed on
            your system, which are required to build this software.
            Please install the xfce4-dev-tools package first, it is available
            from https://www.xfce.org/.
EOF
  exit 1
}

test -d m4 || mkdir m4

XDT_AUTOGEN_REQUIRED_VERSION="4.12.0" \
exec xdt-autogen $@


panel-plugin/usermon.c
/*
 *  Copyright 2003-2020 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#include <stdio.h>
#include <unistd.h>
#include <utmpx.h>
#include <sys/time.h>
#include <sys/types.h>
#include <pwd.h>
#include <time.h>

#include <gtk/gtk.h>
#include <libnotify/notify.h>
#include <libxfce4util/libxfce4util.h>
#include <libxfce4panel/libxfce4panel.h>

#include "usermon.h"
#include "usermon-dialogs.h"

/* default settings */
#define DEFAULT_MAX_USERS_COUNT	2
#define DEFAULT_USERS_COUNT	1
#define DEFAULT_ALARM_PERIOD	5

UserMonitorPlugin *the_usermon_plugin = NULL;

/* prototypes */
static void xfce_usermon_construct(XfcePanelPlugin * plugin);
static void xfce_usermon_free(XfcePanelPlugin * plugin);
static gboolean xfce_usermon_size_changed(XfcePanelPlugin * plugin, gint size);
static void xfce_usermon_mode_changed(XfcePanelPlugin * plugin,
				      XfcePanelPluginMode mode);
static void xfce_usermon_set_timer(void);

/* define the plugin */
XFCE_PANEL_DEFINE_PLUGIN(UserMonitorPlugin, user_monitor)
static void user_monitor_class_init(UserMonitorPluginClass * klass)
{
	XfcePanelPluginClass *plugin_class;

	g_debug("user_monitor_class_init");
	plugin_class = XFCE_PANEL_PLUGIN_CLASS(klass);
	plugin_class->construct = xfce_usermon_construct;
	plugin_class->free_data = xfce_usermon_free;
	plugin_class->size_changed = xfce_usermon_size_changed;
	plugin_class->about = xfce_usermon_show_about;
	plugin_class->configure_plugin = xfce_usermon_configure_plugin;
	plugin_class->mode_changed = xfce_usermon_mode_changed;
}

static void xfce_usermon_read(UserMonitorPlugin * usermon_plugin)
{
	XfceRc *rc;
	struct passwd *passwd = getpwuid(geteuid());
	gchar *file_name;

	g_debug("xfce_usermon_read");

	/* record the current user */
	if ((passwd != NULL) && (passwd->pw_name != NULL)) {
		usermon_plugin->user_name = g_strdup(passwd->pw_name);
		g_hash_table_insert(usermon_plugin->all_users_list,
				    usermon_plugin->user_name,
				    usermon_plugin->user_name);
		usermon_plugin->users_count = 1;
	}

	/* get the plugin config file location */
	file_name =
	    xfce_panel_plugin_save_location(XFCE_PANEL_PLUGIN(usermon_plugin),
					    TRUE);

	if (G_LIKELY(file_name != NULL)) {
		g_debug("Opening %s", file_name);
		/* open the config file, readonly */
		rc = xfce_rc_simple_open(file_name, TRUE);

		/* cleanup */
		g_free(file_name);

		if (G_LIKELY(rc != NULL)) {
			/* read the settings */
			usermon_plugin->max_users_count =
			    xfce_rc_read_int_entry(rc, "max_users_count",
						   DEFAULT_MAX_USERS_COUNT);
			usermon_plugin->alarm_period =
			    xfce_rc_read_int_entry(rc, "alarm_period",
						   DEFAULT_ALARM_PERIOD);

			/* cleanup */
			xfce_rc_close(rc);
		}
	}
}

static void user_monitor_init(UserMonitorPlugin * usermon_plugin)
{
	GtkOrientation orientation;

	/* make only g_error critical */
	g_log_set_always_fatal(G_LOG_LEVEL_ERROR);

	usermon_plugin->log_file = NULL;
	usermon_plugin->ebox = NULL;
	usermon_plugin->hvbox = NULL;
	usermon_plugin->label = NULL;
	usermon_plugin->all_users_list = NULL;
	usermon_plugin->found_users_list = NULL;
	usermon_plugin->user_name = NULL;
	usermon_plugin->max_users_count = DEFAULT_MAX_USERS_COUNT;
	usermon_plugin->users_count = DEFAULT_USERS_COUNT;
	usermon_plugin->alarm_period = DEFAULT_ALARM_PERIOD;
	usermon_plugin->start_time = time(NULL);
	usermon_plugin->last_alarm_time = 0;

	/* get the current orientation */
	orientation =
	    xfce_panel_plugin_get_orientation(XFCE_PANEL_PLUGIN
					      (usermon_plugin));

	/* create panel widgets */
	usermon_plugin->ebox = gtk_event_box_new();
	gtk_widget_show(usermon_plugin->ebox);

	usermon_plugin->hvbox = gtk_box_new(orientation, 2);
	gtk_widget_show(usermon_plugin->hvbox);
	gtk_container_add(GTK_CONTAINER(usermon_plugin->ebox),
			  usermon_plugin->hvbox);

	usermon_plugin->label = gtk_label_new(_("1 User"));
	gtk_widget_show(usermon_plugin->label);
	gtk_box_pack_start(GTK_BOX(usermon_plugin->hvbox),
			   usermon_plugin->label, FALSE, FALSE,
			   DEFAULT_USERMON_PADDING);

	/* create the users lists */
	usermon_plugin->all_users_list =
	    g_hash_table_new(g_str_hash, g_str_equal);
	usermon_plugin->found_users_list =
	    g_hash_table_new(g_str_hash, g_str_equal);

	the_usermon_plugin = usermon_plugin;
}

static void xfce_usermon_free(XfcePanelPlugin * plugin)
{
	UserMonitorPlugin *usermon_plugin;
	GtkWidget *dialog;

	g_debug("xfce_usermon_free");
	usermon_plugin = XFCE_USERMON_PLUGIN(plugin);

	/* check if the dialog is still open. if so, destroy it */
	dialog = g_object_get_data(G_OBJECT(plugin), "dialog");
	if (G_UNLIKELY(dialog != NULL))
		gtk_widget_destroy(dialog);

	/* destroy the panel widgets */
	gtk_widget_destroy(usermon_plugin->hvbox);

	/* destroy the users lists */
	g_hash_table_destroy(usermon_plugin->all_users_list);
	g_hash_table_destroy(usermon_plugin->found_users_list);

	/* cleanup the settings */
	if (G_LIKELY(usermon_plugin->user_name != NULL))
		g_free(usermon_plugin->user_name);

	/* free the plugin structure */
	g_slice_free(UserMonitorPlugin, usermon_plugin);
	the_usermon_plugin = NULL;

	notify_uninit();
}

static gboolean xfce_usermon_size_changed(XfcePanelPlugin * plugin, gint size)
{
	g_debug("xfce_usermon_size_changed");

	/* FIXME: handle this */

	/* we handled the orientation */
	return TRUE;
}

static void
xfce_usermon_mode_changed(XfcePanelPlugin * plugin, XfcePanelPluginMode mode)
{
	g_debug("xfce_usermon_mode_changed");

	/* FIXME: handle this */

	xfce_usermon_size_changed(plugin, xfce_panel_plugin_get_size(plugin));
}

void xfce_usermon_save(XfcePanelPlugin * plugin)
{
	UserMonitorPlugin *usermon_plugin;
	XfceRc *rc;
	gchar *file;

	g_debug("xfce_usermon_save");
	usermon_plugin = XFCE_USERMON_PLUGIN(plugin);

	/* get the config file location */
	file = xfce_panel_plugin_save_location(plugin, TRUE);

	if (G_UNLIKELY(file == NULL)) {
		g_debug("Failed to open config file");
		return;
	}

	/* open the config file, read/write */
	rc = xfce_rc_simple_open(file, FALSE);
	g_free(file);

	if (G_LIKELY(rc != NULL)) {
		/* save the settings */
		xfce_rc_write_int_entry(rc, "max_users_count",
					usermon_plugin->max_users_count);
		xfce_rc_write_int_entry(rc, "alarm_period",
					usermon_plugin->alarm_period);

		/* close the rc file */
		xfce_rc_close(rc);
		g_debug("Saved settings");
	}
}

static void xfce_usermon_show_notification(NotifyUrgency urgency,
					   gchar * body, gint timeout)
{
	NotifyNotification *notification_popup = NULL;

	notification_popup =
	    notify_notification_new(_("User Monitor"),
				    body,
				    PACKAGE_ICON_DIR "/48x48/apps/usermon.png");

	/* display a notification popup */
	if (notification_popup != NULL) {
		notify_notification_set_timeout(notification_popup, timeout);
		notify_notification_set_urgency(notification_popup, urgency);
		notify_notification_show(notification_popup, NULL);

		g_object_unref(G_OBJECT(notification_popup));
	}
}

static void xfce_usermon_notify_for_login(gpointer key,
					  gpointer value, gpointer user_data)
{
	UserMonitorPlugin *usermon_plugin;

	usermon_plugin = XFCE_USERMON_PLUGIN(user_data);

	if (key != NULL && value != NULL && (usermon_plugin->user_name != key)) {
		NotifyUrgency urgency = NOTIFY_URGENCY_NORMAL;
		gchar *body = g_strdup_printf(_("%s logged in"),
					      (gchar *) key);

		if (usermon_plugin->users_count >
		    usermon_plugin->max_users_count) {
			urgency = NOTIFY_URGENCY_CRITICAL;
		}

		xfce_usermon_show_notification(urgency, body,
					       usermon_plugin->alarm_period *
					       1000);

		g_free(body);
	}
}

static void xfce_usermon_notify_for_logout(gpointer key,
					   gpointer value, gpointer user_data)
{
	UserMonitorPlugin *usermon_plugin;

	usermon_plugin = XFCE_USERMON_PLUGIN(user_data);

	if (key != NULL && value != NULL && (usermon_plugin->user_name != key)) {
		NotifyUrgency urgency = NOTIFY_URGENCY_NORMAL;

		g_debug("xfce_usermon_notify_for_logout %s %s", (gchar *) key,
			(gchar *) value);

		/* is this user still logged in? */
		if (g_hash_table_contains
		    (the_usermon_plugin->found_users_list, key) == FALSE) {
			gchar *body = g_strdup_printf(_("%s logged out"),
						      (gchar *) key);

			xfce_usermon_show_notification(urgency, body,
						       usermon_plugin->
						       alarm_period * 1000);

			g_free(body);
		}
	}
}

static void xfce_usermon_record_user(gpointer key,
				     gpointer value, gpointer user_data)
{
	UserMonitorPlugin *usermon_plugin;

	usermon_plugin = XFCE_USERMON_PLUGIN(user_data);

	if (key != NULL && value != NULL) {
		gchar *user_name = g_strdup(key);

		g_debug("xfce_usermon_record_user %s %s", (gchar *) key,
			(gchar *) value);
		g_hash_table_insert(usermon_plugin->all_users_list, user_name,
				    user_name);
	}
}

static void xfce_usermon_free_user(gpointer key,
				   gpointer value, gpointer user_data)
{
	UserMonitorPlugin *usermon_plugin;

	usermon_plugin = XFCE_USERMON_PLUGIN(user_data);

	if (key != NULL && value != NULL && (usermon_plugin->user_name != key)) {
		/* keys are also values */
		g_debug("xfce_usermon_free_user %s %s", (gchar *) key,
			(gchar *) value);
		g_free(key);
	}
}

static void xfce_usermon_update_users_list(void)
{
	GHashTable *new_users_list = NULL;
	struct utmpx *u = NULL;
	guint users_count = 0;

	g_debug("xfce_usermon_update_users_list");
	if (the_usermon_plugin == NULL) {
		return;
	}
	g_debug("Found %d users so far", the_usermon_plugin->users_count);

	new_users_list = g_hash_table_new(g_str_hash, g_str_equal);

	g_debug("Current user is %s", the_usermon_plugin->user_name);

	/* rewind to the beginning of utmpx */
	setutxent();
	/* read utmp */
	while ((u = getutxent())) {
		gchar *user_name;

		if (u->ut_type != USER_PROCESS) {
			continue;
		}

		user_name = g_strdup(u->ut_user);
		g_hash_table_insert(the_usermon_plugin->found_users_list,
				    user_name, user_name);

		/* is this a new user? */
		if (g_hash_table_contains
		    (the_usermon_plugin->all_users_list, user_name) == FALSE) {
			user_name = g_strdup(u->ut_user);

			g_debug("Found new user %s", user_name);
			g_hash_table_insert(new_users_list, user_name,
					    user_name);
		} else {
			g_debug("Found known user %s", user_name);
		}
	}
	/* close utmpx */
	endutxent();

	/* update the label? */
	users_count = g_hash_table_size(the_usermon_plugin->found_users_list);
	if (the_usermon_plugin->users_count != users_count) {
		gtk_widget_destroy(the_usermon_plugin->label);

		/* if utmp is broken for some reason, we may get 0 users */
		if (users_count <= 1) {
			the_usermon_plugin->label = gtk_label_new(_("1 User"));
		} else {
			gchar *label_text =
			    g_strdup_printf(_("%d Users"), users_count);
			the_usermon_plugin->label = gtk_label_new(label_text);
			g_free(label_text);
		}
		gtk_widget_show(the_usermon_plugin->label);
		gtk_box_pack_start(GTK_BOX(the_usermon_plugin->hvbox),
				   the_usermon_plugin->label, FALSE, FALSE,
				   DEFAULT_USERMON_PADDING);
	}
	g_debug("Found %d new users, %d users in total, max is %d",
		g_hash_table_size(new_users_list),
		users_count, the_usermon_plugin->max_users_count);

	/* check for users who have logged out since the last check */
	g_hash_table_foreach(the_usermon_plugin->all_users_list,
			     xfce_usermon_notify_for_logout,
			     the_usermon_plugin);

	/* free the all users list */
	g_hash_table_foreach(the_usermon_plugin->all_users_list,
			     xfce_usermon_free_user, the_usermon_plugin);
	g_hash_table_remove_all(the_usermon_plugin->all_users_list);

	/* copy the current users list into the all users list */
	g_hash_table_insert(the_usermon_plugin->all_users_list,
			    the_usermon_plugin->user_name,
			    the_usermon_plugin->user_name);
	if (users_count > 0) {
		g_hash_table_foreach(the_usermon_plugin->found_users_list,
				     xfce_usermon_record_user,
				     the_usermon_plugin);
	}
	the_usermon_plugin->users_count =
	    g_hash_table_size(the_usermon_plugin->all_users_list);
	g_debug("Reset users list to %d", the_usermon_plugin->users_count);

	/* notify for each new user */
	if (g_hash_table_size(new_users_list) > 0) {
		g_hash_table_foreach(new_users_list,
				     xfce_usermon_notify_for_login,
				     the_usermon_plugin);

		/* update the last alarm time */
		the_usermon_plugin->last_alarm_time = time(NULL);
	}

	/* free the current users list */
	g_hash_table_foreach(the_usermon_plugin->found_users_list,
			     xfce_usermon_free_user, the_usermon_plugin);
	g_hash_table_remove_all(the_usermon_plugin->found_users_list);
	g_hash_table_destroy(new_users_list);
}

static void xfce_usermon_timer_handler(int sig_num)
{
	struct itimerval new_timer;
	g_debug("xfce_usermon_timer_handler");
	if (sig_num == SIGALRM) {
		xfce_usermon_update_users_list();

		if (getitimer(ITIMER_REAL, &new_timer) == 0) {
			/* rearm the timer? */
			if (new_timer.it_value.tv_sec == 0
			    && new_timer.it_value.tv_usec == 0) {
				xfce_usermon_set_timer();
			}
		}
	}
}

static void xfce_usermon_set_timer(void)
{
	struct itimerval new_timer;
	struct sigaction new_action;

	g_debug("xfce_usermon_set_timer");
	memset(&new_action, 0, sizeof(new_action));
	new_action.sa_handler = xfce_usermon_timer_handler;
	sigemptyset(&new_action.sa_mask);
	new_action.sa_flags = 0;

	sigaction(SIGALRM, NULL, NULL);
	if (sigaction(SIGALRM, &new_action, NULL) != 0) {
		g_debug("Failed to set alarm");
		return;
	}

	new_timer.it_value.tv_sec = DEFAULT_ALARM_PERIOD;
	new_timer.it_value.tv_usec = 0;
	new_timer.it_interval.tv_sec = DEFAULT_ALARM_PERIOD;
	new_timer.it_interval.tv_usec = 0;

	if (setitimer(ITIMER_REAL, &new_timer, NULL) != 0) {
		g_debug("Failed to set timer");
	}
}

static void
xfce_usermon_log_handler(const gchar * domain,
			 GLogLevelFlags level,
			 const gchar * message, gpointer data)
{
	UserMonitorPlugin *usermon_plugin = XFCE_USERMON_PLUGIN(data);
	gchar *path;
	const gchar *prefix;

	if (usermon_plugin->log_file == NULL) {
		g_mkdir_with_parents(g_get_user_cache_dir(), 0755);
		path =
		    g_build_filename(g_get_user_cache_dir(),
				     "xfce4-usermon_plugin-plugin.log", NULL);
		usermon_plugin->log_file = fopen(path, "w");
		g_free(path);
	}

	if (usermon_plugin->log_file) {
		switch (level & G_LOG_LEVEL_MASK) {
		case G_LOG_LEVEL_ERROR:
			prefix = "ERROR";
			break;
		case G_LOG_LEVEL_CRITICAL:
			prefix = "CRITICAL";
			break;
		case G_LOG_LEVEL_WARNING:
			prefix = "WARNING";
			break;
		case G_LOG_LEVEL_MESSAGE:
			prefix = "MESSAGE";
			break;
		case G_LOG_LEVEL_INFO:
			prefix = "INFO";
			break;
		case G_LOG_LEVEL_DEBUG:
#ifdef DEBUG
			prefix = "DEBUG";
			break;
#else
			return;
#endif
		default:
			prefix = "LOG";
			break;
		}

		fprintf(usermon_plugin->log_file, "%-10s %-25s %s\n", prefix,
			domain, message);
		fflush(usermon_plugin->log_file);
	}

	/* print log to stdout */
	if (level & G_LOG_LEVEL_ERROR || level & G_LOG_LEVEL_CRITICAL) {
		g_log_default_handler(domain, level, message, NULL);
	}
}

static void xfce_usermon_construct(XfcePanelPlugin * plugin)
{
	UserMonitorPlugin *usermon_plugin = XFCE_USERMON_PLUGIN(plugin);

	notify_init(GETTEXT_PACKAGE);

	xfce_panel_plugin_menu_show_configure(plugin);
	xfce_panel_plugin_menu_show_about(plugin);

	/* setup transation domain */
	xfce_textdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR, "UTF-8");

	/* log messages to a file */
	g_log_set_default_handler(xfce_usermon_log_handler, plugin);

	/* init theme/icon stuff */
	gtk_icon_theme_append_search_path(gtk_icon_theme_get_default(),
					  PACKAGE_ICON_DIR);

	/* FIXME: switch to xfconf */
	/* read the user settings */
	xfce_usermon_read(usermon_plugin);

	/* add the ebox to the panel */
	gtk_container_add(GTK_CONTAINER(plugin), usermon_plugin->ebox);

	/* show the panel's right-click menu on this ebox */
	xfce_panel_plugin_add_action_widget(plugin, usermon_plugin->ebox);

	/* connect plugin signals */
	g_signal_connect(G_OBJECT(plugin), "save",
			 G_CALLBACK(xfce_usermon_save), usermon_plugin);

	/* set the timer */
	xfce_usermon_set_timer();
}


panel-plugin/usermon.h
/*
 *  Copyright 2003-2020 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifndef __USER_MONITOR_H__
#define __USER_MONITOR_H__

#include <stdio.h>
#include <time.h>

G_BEGIN_DECLS typedef struct {
	XfcePanelPluginClass __parent__;
} UserMonitorPluginClass;

typedef struct {
	XfcePanelPlugin __parent__;

	FILE *log_file;

	/* panel widgets */
	GtkWidget *ebox;
	GtkWidget *hvbox;
	GtkWidget *label;

	/* users lists */
	GHashTable *all_users_list;
	GHashTable *found_users_list;

	/* settings */
	gchar *user_name;
	guint max_users_count;
	guint users_count;
	guint alarm_period;
	time_t start_time;
	time_t last_alarm_time;
} UserMonitorPlugin;

#define XFCE_TYPE_USERMON_PLUGIN    (user_monitor_get_type ())
#define XFCE_USERMON_PLUGIN(obj)    (G_TYPE_CHECK_INSTANCE_CAST ((obj), XFCE_TYPE_USERMON_PLUGIN, UserMonitorPlugin))
#define XFCE_USERMON_PLUGIN_CLASS(klass)    (G_TYPE_CHECK_CLASS_CAST ((klass), XFCE_TYPE_USERMON_PLUGIN, UserMonitorPluginClass))
#define XFCE_IS_USERMON_PLUGIN(obj)         (G_TYPE_CHECK_INSTANCE_TYPE ((obj), XFCE_TYPE_USERMON_PLUGIN))
#define XFCE_IS_USERMON_PLUGIN_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE ((klass), XFCE_TYPE_USERMON_PLUGIN))
#define XFCE_USERMON_PLUGIN_GET_CLASS(obj)  (G_TYPE_INSTANCE_GET_CLASS ((obj), XFCE_TYPE_USERMON_PLUGIN, UserMonitorPluginClass))

GType user_monitor_get_type(void) G_GNUC_CONST;

void user_monitor_register_type(XfcePanelTypeModule * type_module);

void xfce_usermon_save(XfcePanelPlugin * plugin);

G_END_DECLS
#endif				/* !__USER_MONITOR_H__ */


panel-plugin/usermon-dialogs.c
/*
 *  Copyright 2003-2021 Fabrice Colin
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#include <string.h>

#include <gtk/gtk.h>
#include <libxfce4ui/libxfce4ui.h>
#include <libxfce4panel/libxfce4panel.h>

#include "usermon.h"
#include "usermon-dialogs.h"

/* the website url */
#define PLUGIN_WEBSITE "https://github.com/FabriceColin/usermon"

static void xfce_usermon_configure_response(GtkWidget * dialog,
					    gint response,
					    UserMonitorPlugin * usermon_plugin)
{
	gboolean result;

	if (response == GTK_RESPONSE_HELP) {
		/* show help */
		result =
		    g_spawn_command_line_async("exo-open --launch WebBrowser "
					       PLUGIN_WEBSITE, NULL);

		if (G_UNLIKELY(result == FALSE)) {
			g_warning(_("Unable to open the following URL: %s"),
				  PLUGIN_WEBSITE);
		}
	} else {
		/* remove the dialog data from the plugin */
		g_object_set_data(G_OBJECT(usermon_plugin), "dialog", NULL);

		/* unlock the panel menu */
		xfce_panel_plugin_unblock_menu(XFCE_PANEL_PLUGIN
					       (usermon_plugin));

		/* save the plugin */
		xfce_usermon_save(XFCE_PANEL_PLUGIN(usermon_plugin));

		/* destroy the properties dialog */
		gtk_widget_destroy(dialog);
	}
}

static void xfce_usermon_max_users_count_changed(GtkSpinButton * spin_button,
						 UserMonitorPlugin *
						 usermon_plugin)
{
	usermon_plugin->max_users_count =
	    gtk_spin_button_get_value_as_int(spin_button);
}

static void xfce_usermon_alarm_period_spin_changed(GtkSpinButton * spin_button,
						   UserMonitorPlugin *
						   usermon_plugin)
{
	usermon_plugin->alarm_period =
	    gtk_spin_button_get_value_as_int(spin_button);
}

static GtkWidget *xfce_usermon_create_layout(UserMonitorPlugin * usermon_plugin)
{
	GtkWidget *vbox =
	    gtk_box_new(GTK_ORIENTATION_VERTICAL, DEFAULT_USERMON_PADDING);
	GtkWidget *row1 =
	    gtk_box_new(GTK_ORIENTATION_HORIZONTAL, DEFAULT_USERMON_PADDING);
	GtkWidget *row2 =
	    gtk_box_new(GTK_ORIENTATION_HORIZONTAL, DEFAULT_USERMON_PADDING);
	GtkWidget *max_users_count_label =
	    gtk_label_new(_("Critical number of users"));
	GtkWidget *max_users_count_spin =
	    gtk_spin_button_new_with_range(1, 100, 1);
	GtkWidget *alarm_period_label = gtk_label_new(_("Alarm period"));
	GtkWidget *alarm_period_spin =
	    gtk_spin_button_new_with_range(5, 3600, 5);
	GtkWidget *alarm_period_label_post = gtk_label_new(_("seconds"));

	gtk_box_pack_start(GTK_BOX(row1), max_users_count_label,
			TRUE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(row1), GTK_WIDGET(max_users_count_spin),
			TRUE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(vbox), row1,
			FALSE, FALSE, 0);

	gtk_box_pack_start(GTK_BOX(row2), alarm_period_label,
			TRUE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(row2), GTK_WIDGET(alarm_period_spin),
			TRUE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(row2), alarm_period_label_post,
			TRUE, FALSE, 0);
	gtk_box_pack_start(GTK_BOX(vbox), row2,
			FALSE, FALSE, 0);

	gtk_spin_button_set_value(GTK_SPIN_BUTTON(max_users_count_spin),
				  (gdouble) usermon_plugin->max_users_count);
	g_signal_connect(G_OBJECT(max_users_count_spin), "value-changed",
			 G_CALLBACK(xfce_usermon_max_users_count_changed),
			 usermon_plugin);

	gtk_spin_button_set_value(GTK_SPIN_BUTTON(alarm_period_spin),
				  (gdouble) usermon_plugin->alarm_period);
	g_signal_connect(G_OBJECT(alarm_period_spin), "value-changed",
			 G_CALLBACK(xfce_usermon_alarm_period_spin_changed),
			 usermon_plugin);

	gtk_widget_show(max_users_count_label);
	gtk_widget_show(max_users_count_spin);
	gtk_widget_show(row1);
	gtk_widget_show(alarm_period_label);
	gtk_widget_show(alarm_period_spin);
	gtk_widget_show(alarm_period_label_post);
	gtk_widget_show(row2);

	return vbox;
}

void xfce_usermon_configure_plugin(XfcePanelPlugin * plugin)
{
	UserMonitorPlugin *usermon_plugin;
	GtkWidget *dialog;
	GtkWidget *vbox;
	GtkWidget *content;

	usermon_plugin = XFCE_USERMON_PLUGIN(plugin);

	/* block the plugin menu */
	xfce_panel_plugin_block_menu(plugin);

	/* create the dialog */
#if LIBXFCE4UI_CHECK_VERSION(4,14,0)
	dialog = xfce_titled_dialog_new_with_mixed_buttons(_("User Monitor"),
				GTK_WINDOW(gtk_widget_get_toplevel(GTK_WIDGET(plugin))),
				GTK_DIALOG_DESTROY_WITH_PARENT,
				"help-browser", _("_Help"), GTK_RESPONSE_HELP,
				"window-close", _("_Close"), GTK_RESPONSE_OK,
				NULL);
#else
	dialog = xfce_titled_dialog_new_with_buttons(_("User Monitor"),
				GTK_WINDOW(gtk_widget_get_toplevel(GTK_WIDGET(plugin))),
				GTK_DIALOG_DESTROY_WITH_PARENT,
				"gtk-help", GTK_RESPONSE_HELP,
				"gtk-close", GTK_RESPONSE_OK,
				NULL);
#endif

	/* set a subtitle */
	xfce_titled_dialog_set_subtitle(XFCE_TITLED_DIALOG(dialog), _("Preferences"));

	/* populate the dialog */
	vbox = gtk_dialog_get_content_area(GTK_DIALOG(dialog));
	content = xfce_usermon_create_layout(usermon_plugin);
	gtk_box_pack_start(GTK_BOX(vbox), content,
			TRUE, FALSE, DEFAULT_USERMON_PADDING);

	/* center dialog on the screen */
	gtk_window_set_position(GTK_WINDOW(dialog), GTK_WIN_POS_CENTER);

	/* set dialog icon */
	gtk_window_set_icon_name(GTK_WINDOW(dialog), "usermon");

	/* link the dialog to the plugin, so we can destroy it when the plugin
	 * is closed, but the dialog is still open */
	g_object_set_data(G_OBJECT(plugin), "dialog", dialog);

	/* connect the reponse signal to the dialog */
	g_signal_connect(G_OBJECT(dialog), "response",
			G_CALLBACK(xfce_usermon_configure_response),
			usermon_plugin);

	/* show the entire dialog */
	gtk_widget_show(content);
	gtk_widget_show(vbox);
	gtk_widget_show(dialog);
}

void xfce_usermon_show_about(XfcePanelPlugin * plugin)
{
	/* about dialog code. you can use the GtkAboutDialog
	 * or the XfceAboutInfo widget */
	GdkPixbuf *icon;

	const gchar *auth[] = {
		"Fabrice Colin <fabrice.colin@gmail.com>",
		NULL
	};

	icon = xfce_panel_pixbuf_from_source(PACKAGE_NAME, NULL, 48);
	gtk_show_about_dialog(NULL, "logo", icon, "license",
			      xfce_get_license_text(XFCE_LICENSE_TEXT_GPL),
			      "version", PACKAGE_VERSION, "program-name",
			      PACKAGE_NAME, "comments",
			      _("User monitor plugin"), "website",
			      PLUGIN_WEBSITE, "copyright",
			      _("Copyright \xc2\xa9 2003-2021 Fabrice Colin"),
			      "authors", auth, NULL);

	if (icon) {
		g_object_unref(G_OBJECT(icon));
	}
}
